File: v1-php-core/src/Engine/StateTracker.php
<?php declare(strict_types=1);

namespace PhpSaga\Engine;

final class StateTracker {
    /** @var array<int, array{status:string, result:mixed, error:?\Throwable}> */
    private array $state = [];
    /** @var list<int> */
    private array $order = [];

    public function markStarted(int $index): void {
        $this->state[$index] = ['status' => 'started', 'result' => null, 'error' => null];
    }

    public function markCompleted(int $index, mixed $result = null): void {
        $this->state[$index] = ['status' => 'completed', 'result' => $result, 'error' => null];
        $this->order[] = $index;
    }

    public function markFailed(int $index, \Throwable $e): void {
        $this->state[$index] = ['status' => 'failed', 'result' => null, 'error' => $e];
    }

    /**
     * Returns step indices in COMPLETION ORDER — used by CompensationEngine
     * @return list<int>
     */
    public function getCompletedSteps(): array {
        return $this->order;
    }

    public function wasStepCompleted(string $name): bool {
        return ($this->state[$name]['status'] ?? '') === 'completed';
    }

    /**
     * @return ?array{name:string, error:\Throwable}
     */
    public function getFailedStep(): ?array {
        foreach ($this->state as $name => $data) {
            if ($data['status'] === 'failed' && $data['error'] instanceof \Throwable) {
                return ['name' => $name, 'error' => $data['error']];
            }
        }
        return null;
    }

    public function reset(): void {
        $this->state = [];
        $this->order = [];
    }
}

File: v4-python/phpsaga/engine/state_tracker.py
from dataclasses import dataclass
from typing import Any, Literal

@dataclass
class StepState:
    name: str
    status: Literal['pending', 'completed', 'failed']
    result: Any = None
    error: Exception | None = None

class StateTracker:
    def __init__(self) -> None:
        self._state: dict[int, StepState] = {}
        self._order: list[int] = [] # completion order (indices)

    def mark_completed(self, index: int, name: str, result: Any = None) -> None:
        self._state[index] = StepState(name, 'completed', result)
        self._order.append(index)

    def mark_failed(self, index: int, name: str, error: Exception) -> None:
        self._state[index] = StepState(name, 'failed', error=error)

    def get_completed_steps(self) -> list[int]:
        return list(self._order) # in completion order

    def was_completed(self, name: str) -> bool:
        return self._state.get(name, StepState('', 'pending')).status == 'completed'
